### PR TITLE
varlena: support bitflag in bigendian format.

### PR DESCRIPTION
**Pull request by Sasasu**
Postgresql has an endian-dependent format for the bitflag in varlena. currently pgrx only support smallendian. adding bigendian support.

Greenplum uses the bigendian format for all platforms.

----

I will double check if there is some copy&past error tomorrow. :>

### CODE REVIEW

**Comment by thomcc** [comment]
> Greenplum uses the bigendian format for all platforms.

This means `target_endian = "little"` is wrong, no?

**Comment by eeeebbbbrrrr** [comment]
The pgrx project has no facilities for testing on big endian architectures.  I haven't yet validated that the new big endian impls of these macros are correct/identical-to-postgres-impls, but even if they are, we can't accept this simply because we can't test it.

Is this part of a larger effort by Greenplum to offer pgrx support?  If so, I'd like an opportunity to speak to someone in the company.  My email is e_ridge @ tcdi.com

File: pgrx/src/varlena.rs [hunk 1]
```diff
@@ -12,17 +12,10 @@ Use of this source code is governed by the MIT license that can be found in the
...
-    // #define SET_VARSIZE_4B(PTR,len) \
-    // 	(((varattrib_4b *) (PTR))->va_4byte.va_header = (((uint32) (len)) << 2))
-
-    // SAFETY:  A varlena can be safely cast to a varattrib_4b
-    (*ptr.cast::<pg_sys::varattrib_4b>()).va_4byte.as_mut().va_header = (len as u32) << 2;
-}
+#[cfg(target_endian = "big")]
+pub use varlena_bigendian::*;
+#[cfg(target_endian = "little")]
+pub use varlena_littleendian::*;
```
Comment (workingjubilee) [inline pgrx/src/varlena.rs:18]: Given what you said
```suggestion
#[cfg(any(feature = "greenplum7", target_endian = "big"))]
pub use varlena_bigendian::*;
#[cfg(not(any(feature = "greenplum7", target_endian = "big")))]
pub use varlena_littleendian::*;
```

File: pgrx/src/varlena.rs [hunk 2]
```diff
@@ -375,3 +280,192 @@ pub fn rust_byte_slice_to_bytea(slice: &[u8]) -> PgBox<pg_sys::bytea> {
         ))
     }
 }
+
+#[cfg(target_endian = "little")]
```
Comment (workingjubilee) [inline pgrx/src/varlena.rs:284]: There really isn't a benefit to putting the cfg here.

```suggestion
```

File: pgrx/src/varlena.rs [hunk 3]
```diff
@@ -375,3 +280,192 @@ pub fn rust_byte_slice_to_bytea(slice: &[u8]) -> PgBox<pg_sys::bytea> {
...
+    /// #define VARATT_IS_1B_E(PTR) ((((varattrib_1b *) (PTR))->va_header) == 0x01)
+    /// ```
+    #[inline]
+    pub unsafe fn varatt_is_1b_e(ptr: *const pg_sys::varlena) -> bool {
+        let va1b = ptr as *const pg_sys::varattrib_1b;
+        (*va1b).va_header == 0x01
+    }
+}
+
+#[cfg(target_endian = "big")]
```
Comment (workingjubilee) [inline pgrx/src/varlena.rs:378]: There really isn't a benefit to putting the cfg here.

```suggestion
```

**Review by workingjubilee — CHANGES_REQUESTED** [review]
This, of course, highlights a deeper problem, which is that we need to actually sit down and agree on the feature for enabling Greenplum support.

File: pgrx/src/varlena.rs (hunk 1 above)
Comment (eeeebbbbrrrr) [reply pgrx/src/varlena.rs:18]: Just to be clear, @workingjubilee's suggestion here is just for demonstration purposes.  We are not in any position to invent a `greenplum7` feature yet.  Such a thing will require a significant amount of planning, design, and understanding on our part, along with the ability to test it... and test it, apparently, on disparate CPU architectures.

Comment (Sasasu) [reply pgrx/src/varlena.rs:18]: My goal is to add greenplum support, but this PR is not related to greenplum. and this PR did not add the `greenplum7` feature. If this PR gets merged, I will add gp7 support based on this PR.

**Comment by Sasasu** [comment]
> > Greenplum uses the bigendian format for all platforms.
> 
> This means `target_endian = "little"` is wrong, no?

this PR is not related to greenplum. for pgrx `target_endian = "little"` is correct.
but greenplum remove the littleendian code, and always use bigendian bitflag on all platforms for an unknown reason.

this bitflag does not use to calculate, so it is possible to use bigendian bitflag on littleendian platforms.

this PR is adding the support for bitflag in bigendian format. not related to greenplum.

**Comment by workingjubilee** [comment]
I understand that, but we strongly prefer to be able to test changes for validation of their correctness, **especially** with the varlena-reading tests. It is very hard for us to validate changes are correct without a big-endian system to test on, and unfortunately it isn't possible to use Rust's byte-swapping facilities such as [`from_be_bytes`](https://doc.rust-lang.org/std/primitive.u32.html#method.from_be_bytes) and [`to_le_bytes`](https://doc.rust-lang.org/std/primitive.u32.html#method.to_le_bytes) to reimplement these macros, to my understanding of the bit-manipulations. If it were, I would unhesitatingly recommend using an "endian-oblivious" implementation.

Most of all, I don't want to approve this and then find out some other part of our code is busted as hell on big-endian systems. Better for it to fail in some blatantly obvious way... I have many grievances with Postgres but it is surprisingly good at recovering from rampant data corruption and erroring when it enters invalid states. I'd rather someone who tried to use PGRX on a big-endian system encounter an early, spectacular failure, than get far enough for them to induce a more subtle data corruption that Postgres does not detect.

**Comment by Sasasu** [comment]
> Most of all, I don't want to approve this and then find out some other part of our code is busted as hell on big-endian systems. 

this is Ok, there is no modern CPU support big-endian anymore. I agree with you.
I will close this PR after I verified on my machine. there still have a problem in `get_nullable_datum` or `varsize_any_exhdr`.

---

and, just an ask,  Can you accept moving most of functions to cshim? This will introduce some function call overhead, which will slightly reduce performance. this ABI problem is already handled in the C side (by static inline and macro functions)

---

for Greenplum support, there are much more works to do. not only the things mentioned in issue 1179.
I will complete my usage first.

**Comment by eeeebbbbrrrr** [comment]
> this is Ok, there is no modern CPU support big-endian anymore. I agree with you.
I will close this PR after I verified on my machine. there still have a problem in get_nullable_datum or varsize_any_exhdr.

If you're talking about https://github.com/tcdi/pgrx/pull/1184/commits/e3464a1f8c4d5c4926aa5dbf717606079d552c5a, please make a separate PR for that -- it does look like it confused length in bytes with number of elements.

> and, just an ask, Can you accept moving most of functions to cshim? This will introduce some function call overhead, which will slightly reduce performance. this ABI problem is already handled in the C side (by static inline and macro functions)

Most the functions being discussed in PR were in the cshim at one point.  The primary problem with the cshim is the ability to easily cross-compile it.  Some downstream pgrx users enjoy cross compiling from x86_64 to aarch64.  In fact, that's a key feature of PL/Rust.

Internally, we've discussed inventing some very lightweight "C to Rust" transformer that could programmatically port these types of C macros.  Of course, this a) doesn't exist yet, and b) wouldn't cover everything, but I think it's a better direction.  We definitely agree that hand-porting these macros is error-prone.

Generally speaking, any of the Postgres #define macros or static inline functions that aren't central to general pgrx operation -- things that would fine if unavailable when the `cshim` feature is disabled -- are fine candidates for the cshim.  Core support for the varlena type is not one of those things, however.

> for Greenplum support, there are much more works to do. not only the things mentioned in issue 1179.
I will complete my usage first.

We'd like to support Greenplum.  

It's important to recognize that pgrx is developed with an understanding of the official Postgres sources/internals.  Supporting deviations from that will require a lot of effort.

As I mentioned earlier, I'd love an opportunity to talk to someone at Greelplum/vmware about this.  In order for the pgrx core team to agree to support multiple, different products we're going to need some coordination first.  I don't want to encourage you or your team to start sending us Greenplum PRs without that coordination.

**Comment by radarwave** [comment]
@eeeebbbbrrrr Thanks for your comments, it would be great to talk with you, and I believe we discussed Zombodb&Greenplum integration before. As a start, maybe we can discuss this in Greenplum open-source Slack channel with both the engineering and product teams. If needed, we can coordinate a time for the meeting.

Slack: greenplum.slack.com
Channel: #extensions

**Comment by Sasasu** [comment]
> If you're talking about https://github.com/tcdi/pgrx/commit/e3464a1f8c4d5c4926aa5dbf717606079d552c5a, please make a separate PR for that
https://github.com/tcdi/pgrx/pull/1190

> Core support for the varlena type is not one of those things, however.
Ok, let us wait for https://github.com/rust-lang/rust-bindgen/pull/2369

>  I'd love an opportunity to talk to someone at Greelplum/vmware about this.
From my side, this is a 20% time project. there are some people who want to use rust in his project and find pgrx is incompatible with Greenplum. and I can write Rust, and I am helping with his project. I don't know the plan at the company level.

and pgrx is working with Greenplum on my fork. I think we need more tests and start internal usage first (not the company plan). I will close this PR one week after (or you can close this PR at any time).

**Comment by workingjubilee** [comment]
> and pgrx is working with Greenplum on my fork

Is https://github.com/Sasasu/pgrx/commit/9f83dbb16fe49020b05cb735d832324da150a147 the fork you mean?

**Comment by Sasasu** [comment]
> Is https://github.com/Sasasu/pgrx/commit/9f83dbb16fe49020b05cb735d832324da150a147 the fork you mean?

yes, this commit breaks Postgresql support. but add Greenplum support (with cfg = pg12).

**Comment by eeeebbbbrrrr** [comment]
My [comment over there](https://github.com/pgcentralfoundation/pgrx/issues/1179#issuecomment-2088740201) applies here too.

**Pull request closed without merging**